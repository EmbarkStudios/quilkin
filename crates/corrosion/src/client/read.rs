//! Deserialization of changes sent from a corrosion agent

pub use corro_api_types::{QueryEvent, SqliteValue};
use eyre::ContextCompat as _;
use quilkin_types::{AddressKind, Endpoint, IcaoCode, TokenSet};
use serde::{
    Deserialize,
    de::{self, SeqAccess},
};
use std::{collections::BTreeSet, fmt, str::FromStr};

pub trait FromSqlValue: Sized {
    fn from_sql(values: &[SqliteValue]) -> eyre::Result<Self>;
}

#[derive(Debug, PartialEq)]
pub struct ServerRow {
    pub endpoint: Endpoint,
    pub icao: IcaoCode,
    pub tokens: TokenSet,
}

pub fn deserialize_token_set(s: &str) -> eyre::Result<TokenSet> {
    let mut ts = BTreeSet::default();

    let mut tokens = data_encoding::BASE64_NOPAD.decode(s.as_bytes())?;

    if tokens.is_empty() {
        return Ok(TokenSet(ts));
    }

    if tokens[0] & 0x80u8 != 0 {
        let len = (tokens[0] & !0x80) as usize;
        for tok in tokens[1..].chunks_exact(len) {
            ts.insert(tok.to_vec());
        }
    } else if tokens[0] > 1 {
        let mut toks = &tokens[1..];
        for _ in 0..tokens[0] as usize {
            let len = toks[0] as usize;
            eyre::ensure!(
                len <= toks.len() - 1,
                "token length {len} is longer than remaining binary slice"
            );

            ts.insert(toks[1..1 + len].to_vec());
            toks = &toks[1 + len..];
        }
    } else {
        tokens.remove(0);
        ts.insert(tokens);
    }

    Ok(TokenSet(ts))
}

#[inline]
pub fn parse_endpoint(addr: &str) -> eyre::Result<Endpoint> {
    let (addr, port) = addr.rsplit_once(':').context("missing ':'")?;
    let port = port.parse()?;
    if let Some(ip) = addr.strip_prefix('|') {
        let ip = ip.parse()?;
        Ok(Endpoint::new(AddressKind::Ip(ip), port))
    } else {
        Ok(Endpoint::new(AddressKind::Name(addr.to_owned()), port))
    }
}

macro_rules! get_column {
    ($index:expr, $name:literal, $v:expr) => {
        $v.get($index)
            .context(concat!("missing column '", $name, "'"))?
            .as_str()
            .context(concat!("column '", $name, "' is not a string"))?
    };
}

macro_rules! get_json {
    ($name:literal, $conv:expr, $seq:expr) => {{
        let v = $seq
            .next_element()?
            .ok_or(de::Error::missing_field($name))?;
        $conv(v).map_err(de::Error::custom)?
    }};
}

impl FromSqlValue for ServerRow {
    fn from_sql(values: &[SqliteValue]) -> eyre::Result<Self> {
        let endpoint = parse_endpoint(get_column!(0, "endpoint", values))?;
        let icao = get_column!(1, "icao", values).parse()?;
        let tokens = deserialize_token_set(get_column!(2, "tokens", values))?;

        Ok(Self {
            endpoint,
            icao,
            tokens,
        })
    }
}

impl<'de> Deserialize<'de> for ServerRow {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct Visitor;

        impl<'de> de::Visitor<'de> for Visitor {
            type Value = ServerRow;

            fn expecting(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                f.write_str("array of server columns")
            }

            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
            where
                A: SeqAccess<'de>,
            {
                let endpoint = get_json!("endpoint", parse_endpoint, seq);
                let icao = get_json!("icao", IcaoCode::from_str, seq);
                let tokens = get_json!("tokens", deserialize_token_set, seq);

                // Ignore the rest of the elements, if we don't we'll leave
                // the deserializer with tokens that will cause an error
                while let Some(Ignore) = seq.next_element()? {}

                Ok(ServerRow {
                    endpoint,
                    icao,
                    tokens,
                })
            }
        }

        deserializer.deserialize_seq(Visitor)
    }
}

pub struct Ignore;

impl<'de> de::Visitor<'de> for Ignore {
    type Value = Self;

    fn expecting(&self, _formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        Ok(())
    }

    #[inline]
    fn visit_bool<E>(self, _x: bool) -> Result<Self::Value, E> {
        Ok(Self)
    }

    #[inline]
    fn visit_i64<E>(self, _x: i64) -> Result<Self::Value, E> {
        Ok(Self)
    }

    #[inline]
    fn visit_i128<E>(self, _x: i128) -> Result<Self::Value, E> {
        Ok(Self)
    }

    #[inline]
    fn visit_u64<E>(self, _x: u64) -> Result<Self::Value, E> {
        Ok(Self)
    }

    #[inline]
    fn visit_u128<E>(self, _x: u128) -> Result<Self::Value, E> {
        Ok(Self)
    }

    #[inline]
    fn visit_f64<E>(self, _x: f64) -> Result<Self::Value, E> {
        Ok(Self)
    }

    #[inline]
    fn visit_str<E>(self, _s: &str) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        Ok(Self)
    }

    #[inline]
    fn visit_none<E>(self) -> Result<Self::Value, E> {
        Ok(Self)
    }

    #[inline]
    fn visit_some<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: de::Deserializer<'de>,
    {
        Self::deserialize(deserializer)
    }

    #[inline]
    fn visit_newtype_struct<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: de::Deserializer<'de>,
    {
        Self::deserialize(deserializer)
    }

    #[inline]
    fn visit_unit<E>(self) -> Result<Self::Value, E> {
        Ok(Self)
    }

    #[inline]
    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
    where
        A: de::SeqAccess<'de>,
    {
        while let Some(Self) = seq.next_element()? {}
        Ok(Self)
    }

    #[inline]
    fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
    where
        A: de::MapAccess<'de>,
    {
        while let Some((Self, Self)) = map.next_entry()? {}
        Ok(Self)
    }

    #[inline]
    fn visit_bytes<E>(self, _bytes: &[u8]) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        Ok(Self)
    }

    fn visit_enum<A>(self, data: A) -> Result<Self::Value, A::Error>
    where
        A: serde::de::EnumAccess<'de>,
    {
        use serde::de::VariantAccess;
        data.variant::<Self>()?.1.newtype_variant()
    }
}

impl<'de> Deserialize<'de> for Ignore {
    #[inline]
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: de::Deserializer<'de>,
    {
        deserializer.deserialize_ignored_any(Self)
    }
}
