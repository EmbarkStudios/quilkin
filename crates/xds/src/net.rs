/*
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use std::{io, net::SocketAddr};

/// TCP listener for a GRPC service, always binds to the local IPv6 address
pub struct TcpListener {
    inner: std::net::TcpListener,
}

impl TcpListener {
    /// Binds a TCP listener, if `None` is passed, binds to an ephemeral port
    #[inline]
    pub fn bind(port: Option<u16>) -> io::Result<Self> {
        std::net::TcpListener::bind((std::net::Ipv6Addr::UNSPECIFIED, port.unwrap_or_default()))
            .map(|inner| Self { inner })
    }

    /// Retrieves the port the listener is bound to
    #[inline]
    pub fn port(&self) -> u16 {
        self.inner.local_addr().expect("failed to bind").port()
    }

    /// Retrieves the local address the listener is bound to
    #[inline]
    pub fn local_addr(&self) -> SocketAddr {
        self.inner.local_addr().expect("failed to bind")
    }

    #[inline]
    pub fn into_stream(self) -> io::Result<tokio_stream::wrappers::TcpListenerStream> {
        self.inner.set_nonblocking(true)?;
        let tl = tokio::net::TcpListener::from_std(self.inner)?;
        Ok(tokio_stream::wrappers::TcpListenerStream::new(tl))
    }
}

impl From<TcpListener> for std::net::TcpListener {
    #[inline]
    fn from(value: TcpListener) -> Self {
        value.inner
    }
}
