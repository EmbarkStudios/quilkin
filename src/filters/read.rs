/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#[cfg(doc)]
use crate::filters::Filter;
use crate::{
    endpoint::{EndpointAddress, UpstreamEndpoints},
    metadata::DynamicMetadata,
};

/// The input arguments to [`Filter::read`].
#[non_exhaustive]
pub struct ReadContext {
    /// The upstream endpoints that the packet will be forwarded to.
    pub endpoints: UpstreamEndpoints,
    /// The source of the received packet.
    pub source: EndpointAddress,
    /// Contents of the received packet.
    pub contents: Vec<u8>,
    /// Arbitrary values that can be passed from one filter to another.
    pub metadata: DynamicMetadata,
}

impl ReadContext {
    /// Creates a new [`ReadContext`].
    pub fn new(endpoints: UpstreamEndpoints, source: EndpointAddress, contents: Vec<u8>) -> Self {
        Self {
            endpoints,
            source,
            contents,
            metadata: DynamicMetadata::new(),
        }
    }

    /// Creates a new [`ReadContext`] from a given [`ReadResponse`].
    pub fn with_response(source: EndpointAddress, response: ReadResponse) -> Self {
        Self {
            endpoints: response.endpoints,
            source,
            contents: response.contents,
            metadata: response.metadata,
        }
    }
}

impl From<ReadContext> for ReadResponse {
    fn from(ctx: ReadContext) -> Self {
        Self {
            endpoints: ctx.endpoints,
            contents: ctx.contents,
            metadata: ctx.metadata,
        }
    }
}

/// The output of [`Filter::read`].
///
/// New instances are created from [`ReadContext`].
///
/// ```rust
/// # use quilkin::filters::{ReadContext, ReadResponse};
///   fn read(ctx: ReadContext) -> Option<ReadResponse> {
///       Some(ctx.into())
///   }
/// ```
#[derive(Debug)]
#[non_exhaustive]
pub struct ReadResponse {
    /// The upstream endpoints that the packet should be forwarded to.
    pub endpoints: UpstreamEndpoints,
    /// Contents of the packet to be forwarded.
    pub contents: Vec<u8>,
    /// Arbitrary values that can be passed from one filter to another
    pub metadata: DynamicMetadata,
}
