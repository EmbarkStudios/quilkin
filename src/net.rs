/*
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

pub mod cluster;
pub mod endpoint;
pub(crate) mod maxmind_db;
pub(crate) mod xds;

use std::{
    io,
    net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr},
};

use socket2::{Protocol, Socket, Type};
use tokio::{net::ToSocketAddrs, net::UdpSocket};

pub use self::{
    cluster::ClusterMap,
    endpoint::{Endpoint, EndpointAddress},
};

/// returns a UdpSocket with address and port reuse, on Ipv6Addr::UNSPECIFIED
fn socket_with_reuse(port: u16) -> std::io::Result<UdpSocket> {
    socket_with_reuse_and_address((Ipv6Addr::UNSPECIFIED, port).into())
}

fn socket_with_reuse_and_address(addr: SocketAddr) -> std::io::Result<UdpSocket> {
    let domain = match addr {
        SocketAddr::V4(_) => socket2::Domain::IPV4,
        SocketAddr::V6(_) => socket2::Domain::IPV6,
    };

    let sock = Socket::new(domain, Type::DGRAM, Some(Protocol::UDP))?;
    enable_reuse(&sock)?;
    sock.set_nonblocking(true)?;
    if domain == socket2::Domain::IPV6 {
        // be explicit so we can have dual stack sockets.
        sock.set_only_v6(false)?;
    }
    sock.bind(&addr.into())?;
    UdpSocket::from_std(sock.into())
}

#[cfg(not(target_family = "windows"))]
fn enable_reuse(sock: &Socket) -> io::Result<()> {
    sock.set_reuse_port(true)?;
    Ok(())
}

#[cfg(target_family = "windows")]
fn enable_reuse(sock: &Socket) -> io::Result<()> {
    sock.set_reuse_address(true)?;
    Ok(())
}

/// An ipv6 socket that can accept and send data from either a local ipv4 address or ipv6 address
/// with port reuse enabled and only_v6 set to false.
#[derive(Debug)]
pub struct DualStackLocalSocket {
    socket: UdpSocket,
}

impl DualStackLocalSocket {
    pub fn new(port: u16) -> std::io::Result<DualStackLocalSocket> {
        Ok(Self {
            socket: socket_with_reuse(port)?,
        })
    }

    pub fn bind_local(port: u16) -> std::io::Result<DualStackLocalSocket> {
        Ok(Self {
            socket: socket_with_reuse_and_address((Ipv6Addr::LOCALHOST, port).into())?,
        })
    }

    /// Primarily used for testing of ipv4 vs ipv6 addresses.
    pub(crate) fn new_with_address(addr: SocketAddr) -> std::io::Result<DualStackLocalSocket> {
        Ok(Self {
            socket: socket_with_reuse_and_address(addr)?,
        })
    }

    pub async fn recv_from(&self, buf: &mut [u8]) -> io::Result<(usize, SocketAddr)> {
        self.socket.recv_from(buf).await
    }

    pub fn local_ipv4_addr(&self) -> io::Result<SocketAddr> {
        let addr = self.socket.local_addr()?;
        match addr {
            SocketAddr::V4(_) => Ok(addr),
            SocketAddr::V6(_) => Ok((Ipv4Addr::UNSPECIFIED, addr.port()).into()),
        }
    }

    pub fn local_ipv6_addr(&self) -> io::Result<SocketAddr> {
        let addr = self.socket.local_addr()?;
        match addr {
            SocketAddr::V4(v4addr) => Ok(SocketAddr::new(
                IpAddr::V6(v4addr.ip().to_ipv6_mapped()),
                addr.port(),
            )),
            SocketAddr::V6(_) => Ok(addr),
        }
    }

    pub async fn send_to<A: ToSocketAddrs>(&self, buf: &[u8], target: A) -> io::Result<usize> {
        self.socket.send_to(buf, target).await
    }
}

#[cfg(test)]
mod tests {
    use std::{
        net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr},
        time::Duration,
    };

    use tokio::time::timeout;

    use crate::net::endpoint::address::AddressKind;
    use crate::test::{available_addr, AddressType, TestHelper};

    #[tokio::test]
    async fn dual_stack_socket_reusable() {
        let expected = available_addr(&AddressType::Random).await;
        let socket = super::DualStackLocalSocket::new(expected.port()).unwrap();
        let addr = socket.local_ipv4_addr().unwrap();

        match expected {
            SocketAddr::V4(_) => assert_eq!(expected, socket.local_ipv4_addr().unwrap()),
            SocketAddr::V6(_) => assert_eq!(expected, socket.local_ipv6_addr().unwrap()),
        }

        assert_eq!(expected.port(), socket.local_ipv4_addr().unwrap().port());
        assert_eq!(expected.port(), socket.local_ipv6_addr().unwrap().port());

        // should be able to do it a second time, since we are reusing the address.
        let socket = super::DualStackLocalSocket::new(expected.port()).unwrap();

        match expected {
            SocketAddr::V4(_) => assert_eq!(expected, socket.local_ipv4_addr().unwrap()),
            SocketAddr::V6(_) => assert_eq!(expected, socket.local_ipv6_addr().unwrap()),
        }
        assert_eq!(addr.port(), socket.local_ipv4_addr().unwrap().port());
        assert_eq!(addr.port(), socket.local_ipv6_addr().unwrap().port());
    }

    #[tokio::test]
    async fn dual_stack_socket() {
        // Since the TestHelper uses the DualStackSocket, we can use it to test ourselves.
        let mut t = TestHelper::default();

        let echo_addr = t.run_echo_server(&AddressType::Random).await;
        let (mut rx, socket) = t.open_socket_and_recv_multiple_packets().await;

        let msg = "hello";
        let addr = echo_addr.to_socket_addr().await.unwrap();

        socket.send_to(msg.as_bytes(), &addr).await.unwrap();
        assert_eq!(
            msg,
            timeout(Duration::from_secs(5), rx.recv())
                .await
                .expect("should not timeout")
                .unwrap()
        );

        // try again, but from the opposite type of IP Address
        // Proof that a dual stack ipv6 socket can send to both ipv6 and ipv4.
        let ipv4_echo_addr = (Ipv4Addr::UNSPECIFIED, echo_addr.port).into();
        let opp_addr: SocketAddr = match echo_addr.host {
            AddressKind::Ip(ip) => match ip {
                IpAddr::V4(_) => (Ipv6Addr::UNSPECIFIED, echo_addr.port).into(),
                IpAddr::V6(_) => ipv4_echo_addr,
            },
            // we're not testing this, since DNS resolves to IP.
            AddressKind::Name(_) => unreachable!(),
        };

        socket.send_to(msg.as_bytes(), &opp_addr).await.unwrap();
        assert_eq!(
            msg,
            timeout(Duration::from_secs(5), rx.recv())
                .await
                .expect("should not timeout")
                .unwrap()
        );

        // Since all other sockets are actual ipv6 sockets, let's force a test with a real ipv4 socket sending to our dual
        // stack socket.
        let (mut rx, socket) = t.open_ipv4_socket_and_recv_multiple_packets().await;
        socket
            .send_to(msg.as_bytes(), &ipv4_echo_addr)
            .await
            .unwrap();
        assert_eq!(
            msg,
            timeout(Duration::from_secs(5), rx.recv())
                .await
                .expect("should not timeout")
                .unwrap()
        );
    }
}

/// Converts a a socket address to its canonical version.
/// This is just a copy of the method available in std but that is currently
/// nightly only.
pub fn to_canonical(addr: &mut SocketAddr) {
    let ip = match addr.ip() {
        std::net::IpAddr::V6(ip) => {
            if let Some(mapped) = ip.to_ipv4_mapped() {
                std::net::IpAddr::V4(mapped)
            } else {
                std::net::IpAddr::V6(ip)
            }
        }
        addr => addr,
    };

    addr.set_ip(ip);
}
