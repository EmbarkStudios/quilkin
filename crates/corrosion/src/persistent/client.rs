use bytes::Bytes;
use corro_api_types::ExecResponse;
use quilkin_types::IcaoCode;
use std::net::SocketAddr;
use tokio::sync::{mpsc, oneshot};

type ResponseTx = oneshot::Sender<Result<ExecResponse, StreamError>>;

#[derive(thiserror::Error, Debug)]
pub enum StreamError {
    #[error(transparent)]
    Connect(#[from] quinn::ConnectionError),
    #[error(transparent)]
    Write(#[from] quinn::WriteError),
    #[error(transparent)]
    Read(#[from] quinn::ReadError),
    #[error(transparent)]
    ReadExact(#[from] quinn::ReadExactError),
    #[error(transparent)]
    Reset(#[from] quinn::ResetError),
    #[error(transparent)]
    Json(serde_json::Error),
    #[error(
        "expected a chunk of JSON length {} but received {}",
        expected,
        received
    )]
    LengthMismatch { expected: usize, received: usize },
    #[error("stream ended")]
    StreamEnded,
}

use super::LengthReadError as Lre;

impl From<Lre> for StreamError {
    fn from(value: Lre) -> Self {
        match value {
            Lre::Json(json) => Self::Json(json),
            Lre::LengthMismatch { expected, received } => {
                Self::LengthMismatch { expected, received }
            }
            Lre::ReadExact(re) => Self::ReadExact(re),
            Lre::Read(r) => Self::Read(r),
            Lre::StreamEnded => Self::StreamEnded,
        }
    }
}

#[derive(thiserror::Error, Debug)]
pub enum ConnectError {
    #[error(transparent)]
    Connect(#[from] quinn::ConnectError),
    #[error(transparent)]
    Connection(#[from] quinn::ConnectionError),
    #[error(transparent)]
    Creation(#[from] std::io::Error),
    #[error(transparent)]
    Handshake(#[from] super::HandshakeError),
    #[error(transparent)]
    Write(#[from] StreamError),
}

#[derive(thiserror::Error, Debug)]
pub enum TransactionError {
    #[error(transparent)]
    Json(#[from] serde_json::Error),
    #[error(transparent)]
    Stream(#[from] StreamError),
    #[error("the I/O task for this client was shutdown")]
    TaskShutdown,
}

/// The current version of the client stream
///
/// - 0: Invalid
/// - 1: The initial version
///   Requests are 16-bit length-prefixed JSON, where the JSON is [`ServerChange`]
///   Responses are the JSON of [`ExecResult`]
pub const VERSION: u16 = 1;

/// A persistent connection to a corrosion agent
pub struct Client {
    inner: quinn::Connection,
    local_addr: SocketAddr,
    tx: mpsc::UnboundedSender<(Bytes, ResponseTx)>,
    task: tokio::task::JoinHandle<Result<Option<quinn::VarInt>, StreamError>>,
}

impl Client {
    /// Connects using a non-encrypted session
    pub async fn connect_insecure(
        addr: SocketAddr,
        qcmp_port: u16,
        icao: IcaoCode,
    ) -> Result<Self, ConnectError> {
        let ep = quinn::Endpoint::client((std::net::Ipv6Addr::LOCALHOST, 0).into())?;

        let inner = ep
            .connect_with(
                quinn_plaintext::client_config(),
                addr,
                &addr.ip().to_string(),
            )?
            .await?;

        // This is really infallible
        let local_addr = ep.local_addr()?;

        let client = inner.clone();
        let (mut send, mut recv) = client.open_bi().await?;

        // Handshake
        // We need to actually send something for the connection to be fully established
        let peer_version = {
            let req = super::ClientHandshakeRequestV1 { qcmp_port, icao }.write();

            send.write_chunk(super::write_length_prefixed(&req).freeze())
                .await
                .map_err(StreamError::from)?;

            let res = super::read_length_prefixed(&mut recv)
                .await
                .map_err(StreamError::from)?;
            match super::ServerHandshake::read(VERSION, &res[..])? {
                super::ServerHandshake::V1(shs) => {
                    if !shs.accept {
                        return Err(ConnectError::Handshake(
                            crate::persistent::HandshakeError::UnsupportedVersion {
                                ours: VERSION,
                                theirs: 1,
                            },
                        ));
                    }

                    1
                }
            }
        };

        let (tx, mut reqrx) = mpsc::unbounded_channel();

        let task = tokio::task::spawn(async move {
            let func = async || -> Result<Option<quinn::VarInt>, StreamError> {
                match peer_version {
                    1 => loop {
                        let (msg, comp): (_, ResponseTx) = tokio::select! {
                            res = recv.received_reset() => {
                                return res.map_err(StreamError::Reset);
                            }
                            req = reqrx.recv() => {
                                let Some(req) = req else {
                                    let _ = send.reset(quinn::VarInt::from_u32(1));
                                    let _ = send.finish();
                                    // We need to drop the recv stream so that the server
                                    // knows we don't care and it can finish closing the connection
                                    drop(recv);
                                    tracing::debug!("waiting for server to received buffered stream...");
                                    drop(send.stopped().await);
                                    tracing::debug!("client finished");
                                    break;
                                };

                                req
                            }
                        };

                        send.write_chunk(msg).await?;
                        let res = super::read_length_prefixed_jsonb::<ExecResponse>(&mut recv)
                            .await
                            .map_err(StreamError::from);

                        if let Err(error) = &res {
                            tracing::error!(%error, "error occurred reading response to transaction");
                        }

                        if comp.send(res).is_err() {
                            tracing::warn!("transaction response could not be sent to queuer");
                        }
                    },
                    _invalid => {
                        return Err(StreamError::Connect(
                            quinn::ConnectionError::VersionMismatch,
                        ));
                    }
                }

                Ok(None)
            };

            func().await
        });

        Ok(Self {
            inner,
            tx,
            task,
            local_addr,
        })
    }

    pub fn local_addr(&self) -> SocketAddr {
        self.local_addr
    }

    pub fn remote_addr(&self) -> SocketAddr {
        self.inner.remote_address()
    }

    pub async fn transactions(
        &self,
        change: &[super::ServerChange],
    ) -> Result<ExecResponse, TransactionError> {
        let buf = super::write_length_prefixed_jsonb(&change)?;

        let (tx, rx) = oneshot::channel();
        self.tx
            .send((buf.freeze(), tx))
            .map_err(|_| TransactionError::TaskShutdown)?;
        Ok(rx.await.map_err(|_| TransactionError::TaskShutdown)??)
    }

    /// Closes the connection to the upstream server
    pub async fn shutdown(self) {
        drop(self.tx);
        if let Ok(Err(error)) = self.task.await {
            tracing::warn!(%error, "stream exited with error");
        }
        drop(self.inner);
    }
}
