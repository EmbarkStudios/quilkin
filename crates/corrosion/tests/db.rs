//! Tests the DB mutations and queries work as expected

use corro_api_types::SqliteValue;
use corro_types::{agent::SplitPool, api::Statement};
use corrosion::client::read::{FromSqlValue, ServerRow};
use corrosion_utils as tu;
use quilkin_types::{AddressKind, Endpoint, IcaoCode};
use std::net::{Ipv4Addr, Ipv6Addr, SocketAddrV6};

async fn exec_all<const N: usize>(v: &mut smallvec::SmallVec<[Statement; N]>, sp: &SplitPool) {
    let mut conn = sp.write_priority().await.unwrap();
    let tx = conn.transaction().unwrap();
    tu::exec(&tx, v.iter()).unwrap();
    tx.commit().unwrap();
    v.clear();
}

async fn read_server_row(id: usize, sp: &SplitPool) -> ServerRow {
    let conn = sp.read().await.unwrap();
    conn.query_row(
        "SELECT endpoint,icao,tokens FROM servers WHERE rowid = ?",
        [id],
        |row| {
            let mut v = Vec::with_capacity(3);
            v.push(row.get::<_, SqliteValue>(0).unwrap());
            v.push(row.get::<_, SqliteValue>(1).unwrap());
            v.push(row.get::<_, SqliteValue>(2).unwrap());
            Ok(ServerRow::from_sql(&v).unwrap())
        },
    )
    .unwrap()
}

fn make_row(i: u32) -> ServerRow {
    let address = match i % 3 {
        0 => AddressKind::Ip(Ipv4Addr::from_bits(i).into()),
        1 => AddressKind::Name(format!("boop.{i}.net")),
        2 => AddressKind::Ip(Ipv6Addr::from_bits(i as _).into()),
        _ => unreachable!(),
    };

    let endpoint = Endpoint {
        address,
        port: i as u16,
    };

    ServerRow {
        endpoint,
        icao: IcaoCode::new_testing([b'B', b'O', b'O', b'P']),
        tokens: [i.to_ne_bytes()].into(),
    }
}

const PREP_PEER: SocketAddrV6 = SocketAddrV6::new(Ipv6Addr::from_bits(0xaaffeeff), 8999, 0, 0);

async fn prep(name: &str) -> SplitPool {
    let sp = tu::new_split_pool(name, corrosion::schema::SCHEMA).await;

    const MAX: usize = 100;

    let mut iv = smallvec::SmallVec::<[_; MAX]>::new();
    let mut s = corrosion::client::write::Server::for_peer(PREP_PEER, &mut iv);

    for i in 0..1000u32 {
        let row = make_row(i);

        if s.statements.len() == MAX {
            exec_all(s.statements, &sp).await;
        }

        s.upsert(&row.endpoint, row.icao, &row.tokens);
    }

    if !s.statements.is_empty() {
        exec_all(s.statements, &sp).await;
    }

    sp
}

/// Tests basics of inserting servers and datacenters
#[tokio::test]
async fn inserts_and_reads_servers() {
    let sp = prep("inserts_and_reads_servers").await;

    {
        let r = sp.read().await.unwrap();
        assert_eq!(
            1000,
            r.query_row("SELECT COUNT(*) FROM servers", [], |r| r.get::<_, u32>(0))
                .unwrap()
        );
    }

    for i in 0..3u32 {
        let row = read_server_row(i as usize + 1, &sp).await;
        let expected = make_row(i);

        assert_eq!(row, expected);
    }
}

/// Tests that servers that have no datacenter contributors are reaped after
/// some amount of time
#[tokio::test]
async fn collects_old_servers() {
    let sp = prep("collects_old_servers").await;
    let fake_time = time::UtcDateTime::now() - std::time::Duration::from_secs(60 * 60);

    {
        let r = sp.read().await.unwrap();
        assert_eq!(
            1000,
            r.query_row("SELECT COUNT(*) FROM servers", [], |r| r.get::<_, u32>(0))
                .unwrap()
        );
    }

    let mut v = smallvec::SmallVec::<[_; 2]>::new();

    // Remove the DC as a contributor, but set the time of the update to an hour in the past
    {
        let mut dc = corrosion::client::write::Datacenter(&mut v);
        dc.remove(PREP_PEER, Some(fake_time));

        let mut conn = sp.write_priority().await.unwrap();
        let tx = conn.transaction().unwrap();
        tu::exec(&tx, dc.0.iter()).unwrap();
        tx.commit().unwrap();
        dc.0.clear();
    }

    {
        let r = sp.read().await.unwrap();
        assert_eq!(
            1000,
            r.query_row("SELECT COUNT(*) FROM servers", [], |r| r.get::<_, u32>(0))
                .unwrap()
        );
    }

    // Add a new server that should not be deleted since it still has a contributor
    {
        let mut s = corrosion::client::write::Server::for_peer(PREP_PEER, &mut v);
        s.upsert(
            &Endpoint {
                address: AddressKind::Ip(Ipv6Addr::from_bits(0x888888888888).into()),
                port: 8888,
            },
            IcaoCode::new_testing([b'V'; 4]),
            &[8888u64.to_ne_bytes()].into(),
        );

        let mut conn = sp.write_priority().await.unwrap();
        let tx = conn.transaction().unwrap();
        tu::exec(&tx, s.statements.iter()).unwrap();
        tx.commit().unwrap();
        s.statements.clear();
    }

    // Do the actual removal of the servers with no contributors that are older than 30 minutes
    {
        let mut s = corrosion::client::write::Server::for_peer(PREP_PEER, &mut v);
        s.reap_old(std::time::Duration::from_secs(60 * 30));

        let mut conn = sp.write_priority().await.unwrap();
        let tx = conn.transaction().unwrap();
        tu::exec(&tx, s.statements.iter()).unwrap();
        tx.commit().unwrap();
    }

    let only_row = {
        let conn = sp.read().await.unwrap();
        let statement = conn
            .prepare("SELECT endpoint,icao,tokens,json(contributors) FROM servers")
            .unwrap();
        tu::query_to_string(statement, |sql, row| {
            row.add_cell(tu::Cell::new(
                &corrosion::client::read::parse_endpoint(&sql.get::<_, String>(0).unwrap())
                    .unwrap()
                    .to_string(),
            ));
            row.add_cell(tu::Cell::new(&sql.get::<_, String>(1).unwrap()));
            row.add_cell(tu::Cell::new(&format!(
                "{:?}",
                corrosion::client::read::deserialize_token_set(&sql.get::<_, String>(2).unwrap())
                    .unwrap()
            )));
            row.add_cell(tu::Cell::new(
                &serde_json::from_str::<serde_json::Value>(&sql.get::<_, String>(3).unwrap())
                    .unwrap()
                    .to_string(),
            ));
        })
    };

    insta::assert_snapshot!("only_one", only_row);
}
