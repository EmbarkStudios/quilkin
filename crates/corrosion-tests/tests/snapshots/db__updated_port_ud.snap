---
source: crates/corrosion/tests/db.rs
expression: only_row().await
---
+-------------+------+------+
| ip          | icao | port |
+=============+======+======+
| ::aaff:eeff | ZZZZ | 9876 |
+-------------+------+------+
